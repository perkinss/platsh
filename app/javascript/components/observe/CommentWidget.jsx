import React, {useState, useEffect, useContext} from 'react'
import { TextField, IconButton, Typography, Tooltip, Grid, Divider } from '@material-ui/core'
import { useTheme } from '@material-ui/styles'
import AddComment from '@material-ui/icons/AddCommentRounded'
import Edit from '@material-ui/icons/Edit'

export default function CommentWidget(props) {
    const { id, comment, handleComment, taskId, studentId, dispatchFormUpdate, classes } = props
    const theme = useTheme()
    const [showField, setShowField] = useState(false)

    const toggleComment = () => {
        setShowField (!showField)
    }

    const handleKeydown = (event) => {
        if (event.which === 13 || event.which === 9 || event.which === 27) {
            setShowField(!showField)
            dispatchFormUpdate({type: 'SET_STUDENT_COMMENT', commentData: {studentId: studentId, taskId: taskId, comment: comment}})
        }
    }

    const handleBlur = (event) => {
        setShowField(!showField)
        dispatchFormUpdate({type: 'SET_STUDENT_COMMENT', commentData: {studentId: studentId, taskId: taskId, comment: comment}})
    }

    const icon = comment ?  <Tooltip title={'Edit comment'}><Edit /></Tooltip> : <Tooltip title={"Add comment"}><AddComment/></Tooltip>

    let displayedContent = null

    if (showField) {
        displayedContent =  <TextField
            id="standard-multiline-flexible"
            label="Comment"
            name={id}
            multiline
            rowsMax="6"
            value={comment}
            onChange={handleComment}
            margin="normal"
            fullWidth={true}
            onKeyDown={handleKeydown}
            autoFocus
            onBlur={handleBlur}
        />
    } else {
        displayedContent =  (
            <React.Fragment>
            {comment && <Tooltip title={comment}><Typography noWrap={true} paragraph onClick={toggleComment}>{comment}</Typography></Tooltip>}
            <IconButton onClick={toggleComment} color={"secondary"} >
                {icon}
            </IconButton>
            </React.Fragment>
        )
    }

    return (

        <Grid container alignItems={'flex-start'} style={{textAlign: 'left'}}>
            <Divider />
            <Grid item xs={12} xl={12}>

            {displayedContent}
            </Grid>
        </Grid>
    )

}