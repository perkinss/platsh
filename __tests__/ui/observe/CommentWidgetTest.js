import React from 'react'
import CommentWidget from '../../../app/javascript/components/observe/CommentWidget'
import { createRender } from '@material-ui/core/test-utils';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import expect from 'expect'

Enzyme.configure({ adapter: new Adapter() });

describe('<CommentWidget />', () => {
    let render;

    beforeAll(() => {
        render = createRender();
    })

    it('should match the snapshot', () => {
        const wrapper = render(<CommentWidget id={1} />);
        expect(wrapper).not.toBeNull()
        expect(wrapper).toMatchSnapshot()
    });

})