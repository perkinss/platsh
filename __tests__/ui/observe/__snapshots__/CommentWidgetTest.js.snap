// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CommentWidget /> should match the snapshot 1`] = `
initialize {
  "0": Node {
    "attribs": Object {
      "class": "MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-flex-start",
      "style": "text-align:left",
    },
    "children": Array [
      Node {
        "attribs": Object {
          "class": "MuiDivider-root",
        },
        "children": Array [],
        "name": "hr",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Node {
          "attribs": Object {
            "class": "MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-xl-12",
          },
          "children": Array [
            Node {
              "attribs": Object {
                "class": "MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary",
                "tabindex": "0",
                "type": "button",
              },
              "children": Array [
                Node {
                  "attribs": Object {
                    "class": "MuiIconButton-label",
                  },
                  "children": Array [
                    Node {
                      "attribs": Object {
                        "aria-hidden": "true",
                        "class": "MuiSvgIcon-root",
                        "focusable": "false",
                        "title": "Add comment",
                        "viewBox": "0 0 24 24",
                      },
                      "children": Array [
                        Node {
                          "attribs": Object {
                            "d": "M22 4c0-1.1-.9-2-2-2H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h14l4 4V4zm-6 7h-3v3c0 .55-.45 1-1 1s-1-.45-1-1v-3H8c-.55 0-1-.45-1-1s.45-1 1-1h3V6c0-.55.45-1 1-1s1 .45 1 1v3h3c.55 0 1 .45 1 1s-.45 1-1 1z",
                          },
                          "children": Array [],
                          "name": "path",
                          "namespace": "http://www.w3.org/2000/svg",
                          "next": null,
                          "parent": [Circular],
                          "prev": null,
                          "type": "tag",
                          "x-attribsNamespace": Object {
                            "d": undefined,
                          },
                          "x-attribsPrefix": Object {
                            "d": undefined,
                          },
                        },
                      ],
                      "name": "svg",
                      "namespace": "http://www.w3.org/2000/svg",
                      "next": null,
                      "parent": [Circular],
                      "prev": null,
                      "type": "tag",
                      "x-attribsNamespace": Object {
                        "aria-hidden": undefined,
                        "class": undefined,
                        "focusable": undefined,
                        "title": undefined,
                        "viewBox": undefined,
                      },
                      "x-attribsPrefix": Object {
                        "aria-hidden": undefined,
                        "class": undefined,
                        "focusable": undefined,
                        "title": undefined,
                        "viewBox": undefined,
                      },
                    },
                  ],
                  "name": "span",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
              ],
              "name": "button",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
                "tabindex": undefined,
                "type": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
                "tabindex": undefined,
                "type": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Node {
        "attribs": Object {
          "class": "MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-xl-12",
        },
        "children": Array [
          Node {
            "attribs": Object {
              "class": "MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary",
              "tabindex": "0",
              "type": "button",
            },
            "children": Array [
              Node {
                "attribs": Object {
                  "class": "MuiIconButton-label",
                },
                "children": Array [
                  Node {
                    "attribs": Object {
                      "aria-hidden": "true",
                      "class": "MuiSvgIcon-root",
                      "focusable": "false",
                      "title": "Add comment",
                      "viewBox": "0 0 24 24",
                    },
                    "children": Array [
                      Node {
                        "attribs": Object {
                          "d": "M22 4c0-1.1-.9-2-2-2H4c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h14l4 4V4zm-6 7h-3v3c0 .55-.45 1-1 1s-1-.45-1-1v-3H8c-.55 0-1-.45-1-1s.45-1 1-1h3V6c0-.55.45-1 1-1s1 .45 1 1v3h3c.55 0 1 .45 1 1s-.45 1-1 1z",
                        },
                        "children": Array [],
                        "name": "path",
                        "namespace": "http://www.w3.org/2000/svg",
                        "next": null,
                        "parent": [Circular],
                        "prev": null,
                        "type": "tag",
                        "x-attribsNamespace": Object {
                          "d": undefined,
                        },
                        "x-attribsPrefix": Object {
                          "d": undefined,
                        },
                      },
                    ],
                    "name": "svg",
                    "namespace": "http://www.w3.org/2000/svg",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "aria-hidden": undefined,
                      "class": undefined,
                      "focusable": undefined,
                      "title": undefined,
                      "viewBox": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "aria-hidden": undefined,
                      "class": undefined,
                      "focusable": undefined,
                      "title": undefined,
                      "viewBox": undefined,
                    },
                  },
                ],
                "name": "span",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
            ],
            "name": "button",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
              "tabindex": undefined,
              "type": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
              "tabindex": undefined,
              "type": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Node {
          "attribs": Object {
            "class": "MuiDivider-root",
          },
          "children": Array [],
          "name": "hr",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
      "style": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
      "style": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
